### PR TITLE
Issue #29303: Add feature to reverse AR/AP Applications

### PR DESCRIPTION
**Pull request by andersdd**
Apply @mccjim scripted package to core

### CODE REVIEW

File: foundation-database/public/functions/reverseapapplication.sql [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  INTO _r
+  FROM apapply 
+  WHERE apapply_id = papplyid;
+
+-- Creates test to not allow to over unapply
+  _paid := (SELECT apopen_paid
+            FROM apopen 
+            WHERE apopen_id = _r.apapply_target_apopen_id);
+
+  IF (_paid < _r.aamt) THEN
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/reverseapapplication.sql]: Rounding and precision errors are possible. Consider something like
```
IF ABS(_paid - _r.aamt) < _roundingError THEN
```

File: foundation-database/public/functions/reverseapapplication.sql [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  WHERE apopen_id = _r.apapply_source_apopen_id;
+
+  UPDATE apopen SET apopen_open = true, apopen_closedate = NULL
+  WHERE apopen_amount != apopen_paid 
+  AND apopen_id = _r.apapply_source_apopen_id;
+
+  UPDATE apopen SET apopen_paid = apopen_paid - _r.aamt 
+  WHERE apopen_id = _r.apapply_target_apopen_id;
+
+  UPDATE apopen SET apopen_open = true, apopen_closedate = NULL
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/reverseapapplication.sql]: This code sequence will fire the trigger on apopen 4 times and looks like it's duplicating some of the trigger logic anyway.

File: foundation-database/public/functions/reverseapapplication.sql [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  IF (_r.apapply_source_doctype = 'K') THEN
+    INSERT INTO apopen(
+            apopen_docdate, apopen_duedate,  apopen_vend_id, 
+            apopen_doctype, apopen_docnumber, apopen_amount, 
+            apopen_notes, 
+             apopen_paid, apopen_open, apopen_username, 
+            apopen_discount, apopen_accnt_id, apopen_curr_id, apopen_curr_rate,
+            apopen_distdate, apopen_void, apopen_discountable_amount, apopen_status)
+    VALUES (current_date, current_date, _r.apapply_vend_id,  
+            'C', _r.apapply_source_docnumber, _r.aamt,
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/reverseapapplication.sql]: `C` vs. `K`?

Won't this second `apopen` record for the check cause problems?

File: foundation-database/public/functions/reverseapapplication.sql [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  FROM apapply 
+  WHERE apapply_id = papplyid;
+
+-- Creates test to not allow to over unapply
+  _paid := (SELECT apopen_paid
+            FROM apopen 
+            WHERE apopen_id = _r.apapply_target_apopen_id);
+
+  IF (_paid < _r.aamt) THEN
+    RAISE EXCEPTION 'Amount paid is greater than the applied amount [xtuple: reverseapapplication, -1]';
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/reverseapapplication.sql]: `greater than` in the message doesn't match `<` in the code

File: foundation-database/public/functions/reverseapapplication.sql (hunk 1 above)
Comment (andersdd) [inline foundation-database/public/functions/reverseapapplication.sql]: This is a specific check to ensure we do not reverse applications greater than the paid amount.

File: foundation-database/public/functions/reverseapapplication.sql (hunk 3 above)
Comment (andersdd) [inline foundation-database/public/functions/reverseapapplication.sql]: I will have to defer to @mccjim as to whether these should be created as credit memos or payments/receipts

**Pull request merged**